### PR TITLE
Add ability to force add / modify Output:SQLite for runs

### PR DESCRIPTION
**Pull request by jmarrec**

### CODE REVIEW

File: epregressions/structures.py [hunk 1]
```diff
@@ -33,6 +33,22 @@ def get_all():
...
+
+class ForceOutputSQLUnitConversion(Enum):
+    NOFORCE = "Don't force anything"
+    NONE = "None"
+    UseOutputControlTableStyle = 'UseOutputControlTableStyle'
+    JtoKWH = 'JtoKWH'
+    JtoMJ = 'JtoMJ'
+    JtoGJ = 'JtoGJ'
+    InchPound = 'InchPound'
+
```
Comment (jmarrec) [inline epregressions/structures.py:50]: new enum classes (cf #89)

File: epregressions/tk_window.py [hunk 2]
```diff
@@ -108,6 +110,10 @@ def __init__(self):
         self.run_period_option.set(ForceRunType.NONE)
         self.reporting_frequency = StringVar()
         self.reporting_frequency.set(ReportingFreq.HOURLY)
+        self.force_output_sql = StringVar()
+        self.force_output_sql.set(ForceOutputSQL.NOFORCE.value)
+        self.force_output_sql_unitconv = StringVar()
+        self.force_output_sql_unitconv.set(ForceOutputSQLUnitConversion.NOFORCE.value)
```
Comment (jmarrec) [inline epregressions/tk_window.py:116]: StringVars to store it. 

(Maybe it'd be better to just use a command with a callback function. Whatever, I'm following current patterns.)

File: epregressions/tk_window.py [hunk 3]
```diff
@@ -233,6 +241,19 @@ def init_window(self):
...
+        self.force_output_sql_option_menu = OptionMenu(
+            group_run_options, self.force_output_sql, *[x.value for x in ForceOutputSQL]
+        )
+        self.force_output_sql_option_menu.grid(row=4, column=2, sticky=W)
+
+        Label(group_run_options, text="Force Output SQL UnitConv: ").grid(row=5, column=1, sticky=E)
+        self.force_output_sql_unitconv_option_menu = OptionMenu(
+            group_run_options, self.force_output_sql_unitconv, *[x.value for x in ForceOutputSQLUnitConversion]
+        )
+        self.force_output_sql_unitconv_option_menu.grid(row=5, column=2, sticky=W)
```
Comment (jmarrec) [inline epregressions/tk_window.py:255]: setup the option menus

File: epregressions/tk_window.py [hunk 4]
```diff
@@ -376,6 +397,9 @@ def client_open(self, auto_open=False):
             self.num_threads_var.set(data['threads'])
             self.run_period_option.set(data['config'])
             self.reporting_frequency.set(data['report_freq'])
+            self.force_output_sql.set(data['force_output_sql'])
+            self.force_output_sql_unitconv.set(data['force_output_sql_unitconv'])
```
Comment (jmarrec) [inline epregressions/tk_window.py:401]: serialize

File: epregressions/tk_window.py [hunk 5]
```diff
@@ -417,6 +441,8 @@ def client_save(self, auto_save=False):
             json_object = {
                 'config': self.run_period_option.get(),
                 'report_freq': self.reporting_frequency.get(),
+                'force_output_sql': self.force_output_sql.get(),
+                'force_output_sql_unitconv': self.force_output_sql_unitconv.get(),
```
Comment (jmarrec) [inline epregressions/tk_window.py:445]: serialize

File: epregressions/tk_window.py [hunk 6]
```diff
@@ -812,6 +838,8 @@ def set_gui_status_for_run(self, is_running: bool):
         self.remove_idf_from_active_button.configure(state=run_button_state)
         self.run_period_option_menu.configure(state=run_button_state)
         self.reporting_frequency_option_menu.configure(state=run_button_state)
+        self.force_output_sql_option_menu.configure(state=run_button_state)
+        self.force_output_sql_unitconv_option_menu.configure(state=run_button_state)
```
Comment (jmarrec) [inline epregressions/tk_window.py:842]: disable when run is launched. 

I think ideally I should also just enable the force_output_sql_unitconv_option_menu only when force_output_sql_option_menu is something else than NOFORCE

File: epregressions/tk_window.py [hunk 7]
```diff
@@ -917,6 +945,8 @@ def client_run(self):
             force_run_type=self.run_period_option.get(),
             num_threads=num_threads,
             report_freq=self.reporting_frequency.get(),
+            force_output_sql=self.force_output_sql.get(),
+            force_output_sql_unitconv=self.force_output_sql_unitconv.get(),
```
Comment (jmarrec) [inline epregressions/tk_window.py:949]: pass to run config

File: epregressions/runtests.py [hunk 8]
```diff
@@ -39,13 +41,17 @@
 
 
 class TestRunConfiguration:
-    def __init__(self, force_run_type, num_threads, report_freq, build_a, build_b, single_test_run=False):
+    def __init__(self, force_run_type, num_threads, report_freq, build_a, build_b, single_test_run=False,
+                 force_output_sql: ForceOutputSQL = ForceOutputSQL.NOFORCE,
+                 force_output_sql_unitconv: ForceOutputSQLUnitConversion = ForceOutputSQLUnitConversion.NOFORCE):
```
Comment (jmarrec) [inline epregressions/runtests.py:46]: Run config ctor: get defaults of NOFORCE.

File: epregressions/runtests.py [hunk 9]
```diff
@@ -39,13 +41,17 @@
...
         self.force_run_type = force_run_type
         self.TestOneFile = single_test_run
         self.num_threads = num_threads
         self.buildA = build_a
         self.buildB = build_b
         self.report_freq = report_freq
+        self.force_output_sql = ForceOutputSQL(force_output_sql)
+        self.force_output_sql_unitconv = ForceOutputSQLUnitConversion(force_output_sql_unitconv)
```
Comment (jmarrec) [inline epregressions/runtests.py:54]: Even if a string is passed, it's coerced into the Enum class

File: epregressions/runtests.py [hunk 10]
```diff
@@ -179,6 +187,50 @@ def read_file_content(file_path):
...
+            if force_output_sql_unitconv == ForceOutputSQLUnitConversion.NOFORCE:
+                idf_text += '\n  Output:SQLite,\n    {};        !- Option Type\n'.format(force_output_sql.value)
+            else:
+                idf_text += '''
+  Output:SQLite,
+    {},        !- Option Type
+    {};        !- Unit Conversion
+    '''.format(force_output_sql.value, force_output_sql_unitconv.value)
+
+        return idf_text
```
Comment (jmarrec) [inline epregressions/runtests.py:232]: staticmethod for adding / modifying Output:SQLite

File: epregressions/runtests.py [hunk 11]
```diff
@@ -316,6 +368,13 @@ def run_build(self, build_tree):
                         )
                 idf_text = idf_text.replace('..\\datasets', 'datasets')
 
+            # Add Output:SQLite if requested
+            if self.force_output_sql != ForceOutputSQL.NOFORCE:
+                idf_text = self.add_or_modify_output_sqlite(
+                    idf_text=idf_text, force_output_sql=self.force_output_sql,
+                    force_output_sql_unitconv=self.force_output_sql_unitconv,
+                )
```
Comment (jmarrec) [inline epregressions/runtests.py:376]: Modify idf_text when requested

File: epregressions/tests/test_runtests.py [hunk 12]
```diff
@@ -2254,3 +2255,107 @@ def test_content_reader(self):
...
+        self.assertEqual("""
+
+  Zone,
+    Zone 1;
+
+  Output:SQLite,
+    SimpleAndTabular,        !- Option Type
+    InchPound;        !- Unit Conversion
+
+""", mod_text)
```
Comment (jmarrec) [inline epregressions/tests/test_runtests.py:2361]: New tests

Comment (Myoldmopar) [reply epregressions/tests/test_runtests.py:2361]: Nice.

**Review by Myoldmopar — COMMENTED** [review]
This seems very cool.  I would love to spend a half day or so back in this codebase and modernize lots of the python.  I do have one maybe silly question - what happens for EpJSON files when you are editing the contents?

**Comment by jmarrec** [comment]
> I do have one maybe silly question - what happens for EpJSON files when you are editing the contents?

That's definitely not a silly question. I didn't even consider (the single) EpJSON to be honest. Fixed + Unit tests in b48ea3176c002da9c6e189d2218a04b6aca6ca67

File: epregressions/tests/test_runtests.py [hunk 13]
```diff
@@ -2254,3 +2255,147 @@ def test_content_reader(self):
...
+        mod_text = SuiteRunner.add_or_modify_output_sqlite(
+            idf_text=idf_text,
+            force_output_sql=ForceOutputSQL.SIMPLEANDTABULAR,
+            force_output_sql_unitconv=ForceOutputSQLUnitConversion.InchPound,
+            isEpJSON=True
+        )
+
+        expected_data = {
+            'Output:SQLite': {'Output:SQLite 4': {'option_type': 'SimpleAndTabular', 'unit_conversion': 'InchPound'}}}
+        self.assertEqual(json.dumps(expected_data, indent=4), mod_text)
```
Comment (jmarrec) [inline epregressions/tests/test_runtests.py:2399]: Unit tests for epJSON

File: epregressions/runtests.py [hunk 14]
```diff
@@ -179,6 +187,67 @@ def read_file_content(file_path):
...
+
+            else:
+                data["Output:SQLite"] = {"Output:SQLite 1": {}}
+                sqlite_obj = data["Output:SQLite"]["Output:SQLite 1"]
+
+            sqlite_obj['option_type'] = force_output_sql.value
+            if force_output_sql_unitconv != ForceOutputSQLUnitConversion.NOFORCE:
+                sqlite_obj['unit_conversion'] = force_output_sql_unitconv.value
+
+            return json.dumps(data, indent=4)
```
Comment (jmarrec) [inline epregressions/runtests.py:216]: New EPSJON handling. It's a json, so read it, do shenanigans, then dump it back to a string.

Comment (Myoldmopar) [reply epregressions/runtests.py:216]: :+1:

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
Spot checked a few files, verified this does force the sql output.  This seems good to go.  And it doesn't have an effect on the regressions if you leave it as don't force anything.  This is just for our own local regression testing.  So yeah, let's merge this.  Thanks @jmarrec !

**Pull request merged**